### ISSUE TITLE
MIEB: Figure out folder structure

### ISSUE THREAD

**Issue by KennethEnevoldsen** [embeddings-benchmark/mteb#1337]
Figure out what to do with folder structure. Should we create a folder like evaluation/evaluators/{modality}, or do we want to do it in another way? I also have some concerns re. evaluators (as I would like them to be the same across text and images, but I see that it might not be possible)

**Comment by isaac-chung** [comment]
Some can certainly be shared without modifications, like the `Any2AnyRetrievalEvaluator`. The rest seem to need some modifications, i.e. unify the modalities from which embeddings are generated. e.g. STSEvaluator and VisualSTSEvaluator share pretty much everything below the line `logger.info("Evaluating...")`.

**Comment by isaac-chung** [comment]
We still have the efforts like creating the MIEB leaderboard (whether separate or aligning the results + adding benchmarks such that it can used in leaderboard v2+). For simplicity, I propose that we:

1. Merge MIEB into main with the current structure, i.e. `evaluation/evaluators/{modality}`
2. Lay out what modifications are needed to reuse evaluators across modalities or create modality-agnostic evaluators / abstasks.
3. Consider how much effort we'd like to spend in extensibility in anticipation of MOEB modalities, e.g. audio.
4. In new issues/PRs, add the new modality-agnostic evaluators / abstasks.
5. When applicable, switch tasks over to use the new modality-agnostic evaluators / abstasks.
6. Mark the older modality-specific evaluators / abstasks as deprecated.

This way, we move this issue to non-blocking, and we can directly start the final merge of mieb into main.

**Comment by KennethEnevoldsen** [comment]
I agree with all of these points. Backward compatibility should be a thing we worry about a lot more after v2 and this seems to align with what you are thinking as well.

wouldn't we want to move text tasks into a "text" folder though?

**Comment by isaac-chung** [comment]
> wouldn't we want to move text tasks into a "text" folder though?

I was hoping the status quo could remain as is, seeing that MTEB was originally set up for `text` only. If we really wanted to do that, maybe v2 would be a good place to do that? I'd say we could leave it out of the MIEB merge to simplify things.

**Comment by x-tabdeveloping** [comment]
I don't think the number of tasks would be a problem, since we can separate benchmarks by modality.

**Comment by Samoed** [comment]
I think we can do:
```
- tasks_classes
  - any_2_any
  - image
  - text
  - AbsTaskClassification
  - AbsTaskSTS
  - ...
 ```
And similar to evaluators and tasks.

**Comment by KennethEnevoldsen** [comment]
@Samoed, how does this look:


```
- abstasks 
  - image # modality-specific implementations
  - text
  - AbsTaskClassification # general-purpose tasks
  - AbsTaskSTS
  - ...
```

**Comment by Samoed** [comment]
I think that's a good! Just make the file names in snake case.

**Comment by isaac-chung** [comment]
I see that evaluators are completed. I think abstasks are completed as well?

**Comment by Samoed** [comment]
No, we haven't change them yet

**Comment by isaac-chung** [comment]
Change the folder structure you mean? There is a separate issue ok snake case so I consider that not within this issue.

**Comment by Samoed** [comment]
Both, we haven't changed folder structure and haven't renamed them. Current folder structure caused by that we've merged some image tasks with multimodal tasks, but we should move some tasks into text folder

**Comment by isaac-chung** [comment]
One last thing before closing this issue is to move the actual tasks to the respective folders.

**Comment by Samoed** [comment]
I think we can have in abstasks
```
image
- image_text_pair_classification
text
- bitext_mining
- pair_classification
- reranking
- summarization
#### "any" tasks ####
- clustering_fast
- multilabel_classification
- aggregated
- classification
- clustering
- sts
- zero shot classification
- regression (not sure if we want)
- retrieval
```

But I'm not how do we want to work with `tasks` folder. We can create move existing tasks to the `text` folder and `image` leave as is

**Comment by isaac-chung** [comment]
Abstasks PR just merged and already follows this structure 👍 

For tasks, we actually should rename the task type folders like we did in abstasks. And I'm open to aligning tasks the exact same way, i.e.

```
tasks
  | - text 
    | - summarization 
      | - eng
        | - taskA
  | - image
  ...
  | - retrieval 
    | - eng
      | - taskB
```

**Comment by Samoed** [comment]
> Abstasks PR just merged and already follows this structure 👍

Not all, clustering fast and multilabel classification in text only folder 

> And I'm open to aligning tasks the exact same way, i.e.

I think we need to see which of retrieval tasks are image and which are texts

**Comment by isaac-chung** [comment]
So this is correct for ClusteringFast as long as this issue is open: https://github.com/embeddings-benchmark/mteb/issues/3320

Moving multi-label will be trivial. I can do that.


> I think we need to see which of retrieval tasks are image and which are texts.

You can check the modalities. But, where would you put a task that is image-text retrieval? Like `it2i`

**Comment by Samoed** [comment]
> You can check the modalities. But, where would you put a task that is image-text retrieval? Like it2i

We can make image and text folder like
```
image
- classification
- image_text_pair_classification
- retrieval (i2i, it2i, i2t, etc)
text
- classification
- - eng
- - - taskA
- bitextmining
- retrieval (only t2t)
```

**Comment by isaac-chung** [comment]
I don't think we should. It2t is not pure image. Instead, we can create a `image-text` folder: 

```
image
- classification (this is i2i only) 
image-text 
text 
```

However, grouping by modalities duplicates the information stored (i.e. types). 

⭐ We should actually group by abstasks in the tasks folder:
1. We should encourage exploring tasks by using `get_tasks()`, which can filter by modalities (image, audio) and types(it2i).
2. We don't duplicate information that's already stored.
